kernel/posix/mqueue.c

```diff
@@ -0,0 +1,532 @@
+
+#define SPACE_PER_MQ (MAX_BLK_SZ +	\
+		      CONFIG_MAX_DESC_PERMQ * sizeof(struct mqueue_desc) +\
+		      sizeof(struct mqueue_object) + sizeof(struct mem_block) +\
+		      CONFIG_MQUEUE_NAMELEN_MAX * sizeof(u8_t))
+
+#define MUL_FACTOR (SPACE_PER_MQ / MAX_BLK_SZ)
+#define CEL_MUL_FACTOR  ((MUL_FACTOR * MAX_BLK_SZ) < SPACE_PER_MQ ?	\
+			  MUL_FACTOR + 1 :	\
+			  MUL_FACTOR)
+K_MEM_POOL_DEFINE(posix_mqueue_pool, MIN_BLK_SZ, MAX_BLK_SZ,
+		  CEL_MUL_FACTOR * CONFIG_MQUEUE_QTY, 4);
+
+K_SEM_DEFINE(mq_sem, 1, 1);
+
+static mqueue_object *mq_list;
+s64_t calculate_timeout(const struct timespec *abstime);
+static mqueue_object *find_in_list(const char *name);
+
+/**
+ * @brief Open a message queue.
+ *
+ * See IEEE 1003.1
+ */
+mqd_t mq_open(const char *name, int oflags, ...)
+{
+	va_list va;
+	mode_t mode;
+	mq_attr *attrs = NULL;
+	u32_t msg_size, max_msgs;
+	mqueue_object  *msg_queue;
+	mqueue_desc *msg_queue_desc = NULL, *mqd = (mqueue_desc *)(-1);
+	u8_t *buffer;
+	struct k_mem_block block_memq_desc, block_memq_obj, block_memq_name;
+	struct k_mem_block block_track, block_buffer;
+	bool error = false;
+
+	va_start(va, oflags);
+	if (oflags & O_CREAT) {
+		mode = va_arg(va, mode_t);
+		attrs = va_arg(va, mq_attr*);
+	}
+	va_end(va);
+
+	if (name == NULL || ((strlen(name) + 1)  > CONFIG_MQUEUE_NAMELEN_MAX)) {
+		errno = EINVAL;
+		return (mqd_t)mqd;
+	}
+
+	/* Check if queue already exists */
+	k_sem_take(&mq_sem, K_FOREVER);
+	msg_queue = find_in_list(name);
+	k_sem_give(&mq_sem);
+
+	if (msg_queue != NULL && oflags & O_CREAT && oflags & O_EXCL) {
+		/* Message queue has alreadey been opened and O_EXCL is set */
+		errno = EEXIST;
+		return (mqd_t)mqd;
+	}
+
+	if (msg_queue == NULL &&
+	    (!(oflags & O_CREAT) || attrs == NULL || attrs->mq_maxmsg <= 0)) {
+		errno = EACCES;
+		return (mqd_t)mqd;
+	}
+
+	/* Allocate message queue descriptor space */
+	if (k_mem_pool_alloc(&posix_mqueue_pool, &block_memq_desc,
+			     sizeof(struct mqueue_desc), 100) == 0) {
+		memset(block_memq_desc.data, 0, sizeof(struct mqueue_desc));
+		msg_queue_desc = block_memq_desc.data;
+		memcpy(&msg_queue_desc->block, &block_memq_desc,
+		       sizeof(struct k_mem_block));
+	} else {
+		errno = ENOSPC;
+		return (mqd_t)mqd;
+	}
+
+	/* Allocate mqueue object for new message queue */
+	if (msg_queue == NULL) {
+
+		/* Check for message quantity and size in message queue */
+		if (attrs->mq_msgsize > CONFIG_MSG_SIZE_MAX &&
+		    attrs->mq_maxmsg > CONFIG_MSG_COUNT_MAX) {
+			error = true;
+			goto done;
+		}
+
+		/* Allocate space for message queue object */
+		if (k_mem_pool_alloc(&posix_mqueue_pool, &block_memq_obj,
+				     sizeof(struct mqueue_object), 100) == 0) {
+			memset(block_memq_obj.data, 0,
+			       sizeof(struct mqueue_object));
+			msg_queue = (struct mqueue_object *)block_memq_obj.data;
+			memcpy(&msg_queue->block_mqueue, &block_memq_obj,
+			       sizeof(struct k_mem_block));
+		} else {
+			error = true;
+			goto done;
```

Comment:
You use this pattern a lot.  Why bother with an "error" boolean when you already have an errno you're preparing to return?  Just check "errno == 0" instead of "!error".

Reply:
hmm good idea. I will incorporate this in new patch set.